### ISSUE TITLE
GetRecentPosts can't deserialize list of posts

### ISSUE THREAD

**Issue by shawnwildermuth** [RickStrahl/MarkdownMonster#257]
Trying to use the Weblog adding and download recent posts but can't seem to de-serialize the post. Is there a way to turn on logging and get a full stack so I can see if it's my libs problem? Here's the error:

![image](https://user-images.githubusercontent.com/568272/31776292-1a01e778-b4b9-11e7-862d-f080a672e818.png)

The full response looks like this, but looking at your RecentPosts structure from the addin seem to look right. Ideas?

```xml
<methodResponse>
  <params>
    <param>
      <value>
        <array>
          <data>
            <value>
              <struct>
                <member>
                  <name>dateCreated</name>
                  <value>
                    <datetime.iso8601>20171019T10:08:37</datetime.iso8601>
                  </value>
                </member>
                <member>
                  <name>description</name>
                  <value>
                    <string>
                      &lt;p&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/Discuss_3.jpg" class="img-fuild pull-right" :="" alt="Tools"&gt;
                      This blog has existed for 15 years now and I've moved it from server to server, service to service, in many forms over the years. As I moved servers, one of my biggest pains was copying all the images and downloads from server to server.&lt;/p&gt;
                      &lt;p&gt;My site code took up about 1% of the space, and all those embedded images and downloads took a majority of the space. I was sick of it, especially on deploying the site (or saving the site in Git), so I decided to switch to storing it in Azure Storage (or AWS if you prefer).&lt;/p&gt;
                      &lt;p&gt;So when I wrote my .NET Core version of the blog, I decided to bite the bullet and start storing them there. But I wanted to enable it directly from Blog authoring. I'm using my version of Metalog API middleware I wrote to do this (see more about that at &lt;a href="https://github.com/shawnwildermuth/metaweblog" target="_blank"&gt;Github&lt;/a&gt;). But I needed a small service to actually support saving new images to the storage service.&lt;/p&gt;
                      &lt;p&gt;First, I set up an Azure Storage account:&lt;/p&gt;
                      &lt;p&gt;&lt;a href="https://wilderminds.blob.core.windows.net/img/azurestorage.jpg" target="_blank"&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/_azurestorage.jpg" class="img-fluid" :="" alt=""&gt;&lt;/a&gt;&lt;/p&gt;
                      &lt;p&gt;Next you'll want to create a container for your images or files in the Azure Storage. A Container is essentially a subdirectory where you can store things. You can have sub-directories under that too.  You'll do this by going &amp;quot;Containers&amp;quot; in your storage account and creating a new container:&lt;/p&gt;
                      &lt;p&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/azurecontainers.png" class="img-fluid" :="" alt=""&gt;&lt;/p&gt;
                      &lt;p&gt;When you create a new container, you can specify what access you want to give. I choose Container to allow anonymous access to the images (so they can be served):&lt;/p&gt;
                      &lt;p&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/azurecontainersecurity.jpg" class="img-fluid" :="" alt=""&gt;&lt;/p&gt;
                      &lt;p&gt;Once you create a storage accont, go to the &amp;quot;Access Keys&amp;quot; and get one of the keys to use to store objects in it:&lt;/p&gt;
                      &lt;p&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/azurekeys.png" class="img-fluid" :="" alt=""&gt;&lt;/p&gt;
                      &lt;p&gt;My first step was getting the Azure Storage library added as a package:&lt;/p&gt;
                      &lt;p&gt;&lt;img src="https://wilderminds.blob.core.windows.net/img/azurepackage.jpg" class="img-fluid" :="" alt=""&gt;&lt;/p&gt;
                      &lt;p&gt;In my project, I started by creating a new service called &lt;strong&gt;ImageStorageService&lt;/strong&gt;. I created a single method StoreImage and refactored an interface to represent my service:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  public class ImageStorageService : IImageStorageService
                      {
                      public async Task&amp;lt;string&amp;gt; StoreImage(string filename, byte[] image)
                      {

                      }
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;The method is asynchronous because I knew that the Azure Storage API was async as well. To use the API, you'll need to construct an URL that matches your Azure URL, Container name, and the new blog (e.g. file) name:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  https://your-storage-name.azure.blob.core.windows.net/your-container-name/new-file-name.jpg
                      // e.g.
                      http://wilderminds.azure.blog.core.windows.net/img/foo.jpg
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;I do this by using path to get the raw filename sent to the method and concatenating it with the URL name. In my case I'm using configuration to store this in a JSON file so I need to inject &lt;strong&gt;IConfiguration&lt;/strong&gt; too:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  public class ImageStorageService : IImageStorageService
                      {
                      private readonly IConfiguration _config;

                      public ImageStorageService(IConfiguration config)
                      {
                      _config = config;
                      }

                      public async Task&amp;lt;string&amp;gt; StoreImage(string filename, byte[] image)
                      {
                      var filenameonly = Path.GetFileName(filename);

                      var url = string.Concat(_config[&amp;quot;BlobService:StorageUrl&amp;quot;], filenameonly);

                      // TODO

                      return url;
                      }
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Next I create storage credentials object by getting the name of the storage account (usually the name of the storage service), and the access key that we copied earlier:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  var creds = new StorageCredentials(_config[&amp;quot;BlobStorage:Account&amp;quot;], _config[&amp;quot;BlobStorage:Key&amp;quot;]);
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Next I create a &lt;strong&gt;CloudBlockBlog&lt;/strong&gt; object to store the file:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  var blob = new CloudBlockBlob(new Uri(url), creds);
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Next I can use the blob object to check and see if it exists and if so, if it is the same size. I do this by first calling ExistsAsync, and then retrieving the attributes via FetchAttributesAsync and checking the length against the length. This makes sure that we don't re-upload the same file over and over:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;      bool shouldUpload = true;
                      if (await blob.ExistsAsync())
                      {
                      await blob.FetchAttributesAsync();
                      if (blob.Properties.Length == image.Length)
                      {
                      shouldUpload = false;
                      }
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Lastly, I send the file via &lt;strong&gt;UploadFromByteArrayAsync&lt;/strong&gt;. If you have a rather large file to upload, you might need to chunk it up, but this call is like a stream so you can send it in pieces:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  if (shouldUpload) await blob.UploadFromByteArrayAsync(image, 0, image.Length);
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;So the complete method looks like this:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  public async Task&amp;lt;string&amp;gt; StoreImage(string filename, byte[] image)
                      {
                      var filenameonly = Path.GetFileName(filename);

                      var url = string.Concat(_config[&amp;quot;BlobService:StorageUrl&amp;quot;], filenameonly);
                      var creds = new StorageCredentials(_config[&amp;quot;BlobStorage:Account&amp;quot;], _config[&amp;quot;BlobStorage:Key&amp;quot;]);
                      var blob = new CloudBlockBlob(new Uri(url), creds);

                      if (!(await blob.ExistsAsync()))
                      {
                      await blob.UploadFromByteArrayAsync(image, 0, image.Length);
                      }

                      return url;
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Now that it's complete, I just register the service in &lt;strong&gt;Startup.cs&lt;/strong&gt;:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;svcs.AddTransient&amp;lt;IImageStorageService, ImageStorageService&amp;gt;();
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;To call this, I just inject the service into my &lt;strong&gt;Metaweblog&lt;/strong&gt; service like so:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;  public class WilderWeblogProvider : IMetaWeblogProvider
                      {
                      private IWilderRepository _repo;
                      private UserManager&amp;lt;WilderUser&amp;gt; _userMgr;
                      private IConfiguration _config;
                      private IHostingEnvironment _appEnv;
                      private readonly IImageStorageService _imageService;

                      public WilderWeblogProvider(UserManager&amp;lt;WilderUser&amp;gt; userMgr,
                      IWilderRepository repo,
                      IConfiguration config,
                      IHostingEnvironment appEnv,
                      IImageStorageService imageService)
                      {
                      _repo = repo;
                      _userMgr = userMgr;
                      _config = config;
                      _appEnv = appEnv;
                      _imageService = imageService;
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;And finally, just call the &lt;strong&gt;StoreImage&lt;/strong&gt; method:&lt;/p&gt;
                      &lt;pre&gt;&lt;code&gt;    public MediaObjectInfo NewMediaObject(string blogid, string username, string password, MediaObject mediaObject)
                      {
                      EnsureUser(username, password).Wait();

                      var bits = Convert.FromBase64String(mediaObject.bits);
                      var op = _imageService.StoreImage(mediaObject.name, bits);

                      op.Wait();
                      if (!op.IsCompletedSuccessfully) throw op.Exception;
                      var url = op.Result;

                      // Create the response
                      MediaObjectInfo objectInfo = new MediaObjectInfo();
                      objectInfo.url = url;

                      return objectInfo;
                      }
                      &lt;/code&gt;&lt;/pre&gt;
                      &lt;p&gt;Note the work involved with the &lt;strong&gt;op.Wait()&lt;/strong&gt; is mostly because the Metaweblog must be synchronous and the method I'm calling is asynchronous, if you use this in a controller, you can juse use &lt;strong&gt;async&lt;/strong&gt; and &lt;strong&gt;await&lt;/strong&gt;.&lt;/p&gt;
                      &lt;p&gt;You could do this via a web upload or even in an API if you needed to. You can see the code on the blog's github:&lt;/p&gt;
                      &lt;blockquote&gt;
                      &lt;p&gt;&lt;a href="https://github.com/shawnwildermuth/wilderblog"&gt;https://github.com/shawnwildermuth/wilderblog&lt;/a&gt;&lt;/p&gt;
                      &lt;/blockquote&gt;
                      &lt;p&gt;Hope this helps!&lt;/p&gt;
                    </string>
                  </value>
                </member>
                <member>
                  <name>title</name>
                  <value>
                    <string>Using Azure Storage in ASP.NET Core</string>
                  </value>
                </member>
                <member>
                  <name>categories</name>
                  <value>
                    <array>
                      <data>
                        <value>
                          <string>Azure</string>
                        </value>
                        <value>
                          <string>ASP.NET Core</string>
                        </value>
                      </data>
                    </array>
                  </value>
                </member>
                <member>
                  <name>permalink</name>
                  <value>
                    <string>2017/10/19/Using-Azure-Storage-in-ASP-NET-Core</string>
                  </value>
                </member>
                <member>
                  <name>postid</name>
                  <value>
                    <i4>1</i4>
                  </value>
                </member>
                <member>
                  <name>userid</name>
                  <value>
                    <string>shawnwildermuth</string>
                  </value>
                </member>
                <member>
                  <name>wp_slug</name>
                  <value>
                    <string>2017/10/19/Using-Azure-Storage-in-ASP-NET-Core</string>
                  </value>
                </member>
              </struct>
            </value>
          </data>
        </array>
      </value>
    </param>
  </params>
</methodResponse>
```

**Comment by RickStrahl** [comment]
Took a look and compared your output to what gets sent back from my blog and few others - it looks fine to me as well, so I'm not quite sure what's going on.

The processing of this happens inside of the XmlRpc Library at the protocol level and the structure returned should be array of XmlRpcStruct[]:

```cs
    [XmlRpcMissingMapping(MappingAction.Ignore)]
    public struct XmlRpcRecentPost
    {
        public DateTime dateCreated;
        public string userid;
        public string description;
        public string title;
        public object postid;
        public string link;
        public string permaLink;
        public string[] categories;

        public string mt_excerpt;
        public string mt_text_more;
        public string wp_more_text;
        public int mt_allow_comments;
        public int mt_allow_pings;
        public string mt_keywords;
        public string wp_slug;
        public string wp_password;
        public string wp_author_id;
        public string wp_author_display_name;
        public string post_status;
        public string post_type;

        public XmlRpcCustomField[] custom_fields;
        public XmlRpcTerm[] terms;

        public override string ToString()
        {
            return this.description;
        }
    }
```

which looks fine for what you have - missing fields should be ignored (and they are as I don't send all of them either), so I don't see anything that's not lining up.

Generally you also want to send back an abstract not the full post because if you have a lot of entries you want to try and keep the list returned to a reasonable size.

Is there a way to create a separate account on your custom blog? If so I can try hooking up to it and see if there's something that I can figure out.

I see errors in the crash logs from you, but they're all related to posting (the bool issue you posted I think ). I don't see anything for GetRecentPosts.



FWIW, here's what I sent back from my blog in case you want to compare (I did but I don't see any glaring difference):

```xml
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value>
        <array>
          <data>
            <value>
              <struct>
                <member>
                  <name>dateCreated</name>
                  <value>
                    <dateTime.iso8601>20171013T14:28:20</dateTime.iso8601>
                  </value>
                </member>
                <member>
                  <name>description</name>
                  <value>
                    <string />
                  </value>
                </member>
                <member>
                  <name>title</name>
                  <value>
                    <string>Opening a Web Browser with an HTTP Url from Visual Studio Code</string>
                  </value>
                </member>
                <member>
                  <name>categories</name>
                  <value>
                    <array>
                      <data>
                        <value>
                          <string>Visual Studio Code</string>
                        </value>
                      </data>
                    </array>
                  </value>
                </member>
                <member>
                  <name>enclosure</name>
                  <value>
                    <struct>
                      <member>
                        <name>length</name>
                        <value>
                          <i4>0</i4>
                        </value>
                      </member>
                    </struct>
                  </value>
                </member>
                <member>
                  <name>link</name>
                  <value>
                    <string>https://weblog.west-wind.com/Opening-a-Web-Browser-with-an-HTTP-Url-from-Visual-Studio-Code</string>
                  </value>
                </member>
                <member>
                  <name>permalink</name>
                  <value>
                    <string>https://weblog.west-wind.com/Opening-a-Web-Browser-with-an-HTTP-Url-from-Visual-Studio-Code</string>
                  </value>
                </member>
                <member>
                  <name>postid</name>
                  <value>
                    <i4>444227</i4>
                  </value>
                </member>
                <member>
                  <name>source</name>
                  <value>
                    <struct />
                  </value>
                </member>
                <member>
                  <name>mt_excerpt</name>
                  <value>
                    <string>Here's a quick tip on how you can create a project specific Visual Studio Task that can launch a Web Browser to a specific page quickly and easily.</string>
                  </value>
                </member>
              </struct>
            </value>
            <value>
              <struct>
                <member>
                  <name>dateCreated</name>
                  <value>
                    <dateTime.iso8601>20170926T09:09:13</dateTime.iso8601>
                  </value>
                </member>
                <member>
                  <name>description</name>
                  <value>
                    <string />
                  </value>
                </member>
                <member>
                  <name>title</name>
                  <value>
                    <string>A few notes on creating Class Libraries for ASP.NET Core</string>
                  </value>
                </member>
                <member>
                  <name>categories</name>
                  <value>
                    <array>
                      <data>
                        <value>
                          <string>ASP.NET Core</string>
                        </value>
                      </data>
                    </array>
                  </value>
                </member>
                <member>
                  <name>enclosure</name>
                  <value>
                    <struct>
                      <member>
                        <name>length</name>
                        <value>
                          <i4>0</i4>
                        </value>
                      </member>
                    </struct>
                  </value>
                </member>
                <member>
                  <name>link</name>
                  <value>
                    <string>https://weblog.west-wind.com/A-few-notes-on-creating-Class-Libraries-for-ASPNET-Core</string>
                  </value>
                </member>
                <member>
                  <name>permalink</name>
                  <value>
                    <string>https://weblog.west-wind.com/A-few-notes-on-creating-Class-Libraries-for-ASPNET-Core</string>
                  </value>
                </member>
                <member>
                  <name>postid</name>
                  <value>
                    <i4>417742</i4>
                  </value>
                </member>
                <member>
                  <name>source</name>
                  <value>
                    <struct />
                  </value>
                </member>
                <member>
                  <name>mt_excerpt</name>
                  <value>
                    <string>I recently started to collect some of my ASP.NET Core utilities into a helper support library as I tend to do. In the process I ran into a few snags and I realized I was making a few non-obvious mistakes right from the start. In this post I discuss a few of the issues with dependencies and how to deal with them.</string>
                  </value>
                </member>
              </struct>
            </value>
            <value>
              <struct>
                <member>
                  <name>dateCreated</name>
                  <value>
                    <dateTime.iso8601>20170925T10:54:30</dateTime.iso8601>
                  </value>
                </member>
                <member>
                  <name>description</name>
                  <value>
                    <string />
                  </value>
                </member>
                <member>
                  <name>title</name>
                  <value>
                    <string>WPF Slow Window Loading due to Invalid SpellChecking Dictionaries</string>
                  </value>
                </member>
                <member>
                  <name>categories</name>
                  <value>
                    <array>
                      <data>
                        <value>
                          <string>WPF</string>
                        </value>
                      </data>
                    </array>
                  </value>
                </member>
                <member>
                  <name>enclosure</name>
                  <value>
                    <struct>
                      <member>
                        <name>length</name>
                        <value>
                          <i4>0</i4>
                        </value>
                      </member>
                    </struct>
                  </value>
                </member>
                <member>
                  <name>link</name>
                  <value>
                    <string>https://weblog.west-wind.com/WPF-Slow-Window-Loading-due-to-Invalid-SpellCheck-Dictionaries</string>
                  </value>
                </member>
                <member>
                  <name>permalink</name>
                  <value>
                    <string>https://weblog.west-wind.com/WPF-Slow-Window-Loading-due-to-Invalid-SpellCheck-Dictionaries</string>
                  </value>
                </member>
                <member>
                  <name>postid</name>
                  <value>
                    <i4>416254</i4>
                  </value>
                </member>
                <member>
                  <name>source</name>
                  <value>
                    <struct />
                  </value>
                </member>
                <member>
                  <name>mt_excerpt</name>
                  <value>
                    <string>I ran into a nasty problem with spell checking on WPF, which caused any form that uses spell checking to load extremely slow. It turns out the problem was caused by errant entries in the global Windows dictionary key in the registry. This post describes the problem and how to find and fix your global Windows dictionary settings.</string>
                  </value>
                </member>
              </struct>
            </value>
            <value>
              <struct>
                <member>
                  <name>dateCreated</name>
                  <value>
                    <dateTime.iso8601>20170918T11:07:57</dateTime.iso8601>
                  </value>
                </member>
                <member>
                  <name>description</name>
                  <value>
                    <string />
                  </value>
                </member>
                <member>
                  <name>title</name>
                  <value>
                    <string>Conditional TargetFrameworks for Multi-Targeted .NET SDK Projects on Cross-Platform Builds</string>
                  </value>
                </member>
                <member>
                  <name>categories</name>
                  <value>
                    <array>
                      <data>
                        <value>
                          <string>.NET</string>
                        </value>
                        <value>
                          <string> ASP.NET Core</string>
                        </value>
                      </data>
                    </array>
                  </value>
                </member>
                <member>
                  <name>enclosure</name>
                  <value>
                    <struct>
                      <member>
                        <name>length</name>
                        <value>
                          <i4>0</i4>
                        </value>
                      </member>
                    </struct>
                  </value>
                </member>
                <member>
                  <name>link</name>
                  <value>
                    <string>https://weblog.west-wind.com/Conditional-TargetFrameworks-for-MultiTargeted-NET-SDK-Projects-on-CrossPlatform-Builds</string>
                  </value>
                </member>
                <member>
                  <name>permalink</name>
                  <value>
                    <string>https://weblog.west-wind.com/Conditional-TargetFrameworks-for-MultiTargeted-NET-SDK-Projects-on-CrossPlatform-Builds</string>
                  </value>
                </member>
                <member>
                  <name>postid</name>
                  <value>
                    <i4>405197</i4>
                  </value>
                </member>
                <member>
                  <name>source</name>
                  <value>
                    <struct />
                  </value>
                </member>
                <member>
                  <name>mt_excerpt</name>
                  <value>
                    <string>If you build multi-targeted .NET SDK projects on multiple platforms you're going to find out that certain targets can't be build on certain platforms. If you target NetStandard and Net45 on a Mac, Net45 is going to fail. In order to get around this you need to conditionally build per platform. Here's how.</string>
                  </value>
                </member>
              </struct>
            </value>
          </data>
        </array>
      </value>
    </param>
  </params>
</methodResponse>
```

**Comment by shawnwildermuth** [comment]
Thanks, let me see what I find out.

**Comment by shawnwildermuth** [comment]
I can't seem to figure out anything different. Any way to turn on logging so I can see the entire exception (and maybe log the request it's receiving)?

**Comment by RickStrahl** [comment]
Yeah I don't know either offhand - your XML looks fine from what I can tell. 

There's really no additional information there to give you - it comes from the XML RPC stack.

If you want to build from source, you can look at:

https://github.com/RickStrahl/MarkdownMonster/blob/master/AddIns/WebLogAddin/MetaWebLogApi/wrappers/MetaWeblogWrapper.cs#L78

You can see the raw RPC excetion there and see if there's additional info.

**Comment by shawnwildermuth** [comment]
Will do.

**Comment by RickStrahl** [comment]
So I finally had some time to look at this a little closer. 

I set up a test request that returns your raw data and I see the same result. After a quite a bit of experimenting I think I found the problem: 

Your date isn't working, although I'm not sure exactly why. The array error probably comes from the date structure internally used to serialize/deserialize the date and something in the date fails.

This fails:

```xml
<member>
                  <name>dateCreated</name>
                  <value>
                    <datetime.iso8601>20171019T10:08:37</datetime.iso8601>
                  </value>
</member>
```

This works:

```xml
<member>
                  <name>dateCreated</name>
                  <value>
                    <dateTime.iso8601>20171013T14:28:20</dateTime.iso8601>
                  </value>
</member>
```

Looks like the difference is case on `dateTime` - you're using `datetime` on the value.

![image](https://user-images.githubusercontent.com/1374013/32410049-543daf30-c15c-11e7-80df-3cd181388884.png)
